pinot-core/src/main/java/org/apache/pinot/core/segment/processing/framework/SegmentProcessorFramework.java

```diff
@@ -0,0 +1,194 @@
+      throw new IllegalStateException("No segments found in input dir: " + _inputSegmentsDir.getAbsolutePath()
+          + ". Exiting SegmentProcessorFramework.");
+    }
+
+    // Mapper phase.
+    LOGGER.info("Beginning mapper phase. Processing segments: {}", Arrays.toString(_inputSegmentsDir.list()));
+    for (File segment : segmentFiles) {
+
+      String fileName = segment.getName();
+      File mapperInput = segment;
+
+      // Untar the segments if needed
+      if (!segment.isDirectory()) {
+        if (fileName.endsWith(".tar.gz") || fileName.endsWith(".tgz")) {
+          mapperInput = TarGzCompressionUtils.untar(segment, _mapperInputDir).get(0);
+        } else {
+          throw new IllegalStateException("Unsupported segment format: " + segment.getAbsolutePath());
+        }
+      }
+
+      // Set mapperId as the name of the segment
+      SegmentMapperConfig mapperConfig =
+          new SegmentMapperConfig(_pinotSchema, _segmentProcessorConfig.getRecordTransformerConfig(),
+              _segmentProcessorConfig.getRecordFilterConfig(), _segmentProcessorConfig.getPartitioningConfig());
+      SegmentMapper mapper = new SegmentMapper(mapperInput.getName(), mapperInput, mapperConfig, _mapperOutputDir);
+      mapper.map();
+      mapper.cleanup();
+    }
+
+    // Check for mapper output files
+    File[] mapperOutputFiles = _mapperOutputDir.listFiles();
+    if (mapperOutputFiles.length == 0) {
+      throw new IllegalStateException("No files found in mapper output directory: " + _mapperOutputDir.getAbsolutePath()
+          + ". Exiting SegmentProcessorFramework.");
+    }
+
+    // Reducer phase.
+    LOGGER.info("Beginning reducer phase. Processing files: {}", Arrays.toString(_mapperOutputDir.list()));
+    // Mapper output directory has 1 directory per partition, named after the partition. Each directory contains 1 or more avro files.
+    for (File partDir : mapperOutputFiles) {
+
+      // Set partition as reducerId
+      SegmentReducerConfig reducerConfig =
+          new SegmentReducerConfig(_pinotSchema, _segmentProcessorConfig.getCollectorConfig(),
+              _segmentProcessorConfig.getSegmentConfig().getMaxNumRecordsPerSegment());
+      SegmentReducer reducer = new SegmentReducer(partDir.getName(), partDir, reducerConfig, _reducerOutputDir);
+      reducer.reduce();
+      reducer.cleanup();
+    }
+
+    // Check for reducer output files
+    File[] reducerOutputFiles = _reducerOutputDir.listFiles();
+    if (reducerOutputFiles.length == 0) {
+      throw new IllegalStateException(
+          "No files found in reducer output directory: " + _reducerOutputDir.getAbsolutePath()
+              + ". Exiting SegmentProcessorFramework.");
+    }
+
+    // Segment generation phase.
+    LOGGER.info("Beginning segment generation phase. Processing files: {}", Arrays.toString(_reducerOutputDir.list()));
+    // Reducer output directory will have 1 or more avro files
+    for (File resultFile : reducerOutputFiles) {
```

Comment:
Did you check the output segment names when the output is more than 1 files?

It's possible that the final segments may have the same segment name. (e.g. `<tablename>_<start>_<end>`)

Reply:
Good catch. Will add seqId. Also working on adding end-to-end tests for the framework, and will include this case